presto-main/src/main/java/com/facebook/presto/metadata/FunctionManager.java

```diff
@@ -0,0 +1,224 @@
+    private final FunctionNamespace operatorNamespace;
+    private final TypeManager typeManager;
+    private final BlockEncodingSerde blockEncodingSerde;
+    private final FeaturesConfig featuresConfig;
+
+    public FunctionManager(TypeManager typeManager, BlockEncodingSerde blockEncodingSerde, FeaturesConfig featuresConfig)
+    {
+        functionNamespaces = ImmutableMap.of();
+        this.typeManager = requireNonNull(typeManager, "typeManager is null");
+        this.blockEncodingSerde = requireNonNull(blockEncodingSerde, "blockEncodingSerde is null");
+        this.featuresConfig = requireNonNull(featuresConfig, "featuresConfig is null");
+        operatorNamespace = new StaticFunctionNamespace(typeManager, blockEncodingSerde, featuresConfig);
+    }
+
+    public void addFunctions(List<? extends SqlFunction> functions)
+    {
+        addFunctions(TEMP_DEFAULT_CATALOG, functions);
+    }
+
+    public synchronized void addFunctions(String catalog, List<? extends SqlFunction> functions)
+    {
+        Map<Boolean, List<SqlFunction>> operatorAndFunctionMap = functions.stream()
+                .collect(Collectors.partitioningBy(function -> isOperator(function.getSignature())));
+
+        operatorNamespace.addFunctions(operatorAndFunctionMap.get(true));
+
+        if (!functionNamespaces.containsKey(catalog)) {
+            FunctionNamespace namespace = new StaticFunctionNamespace(typeManager, blockEncodingSerde, featuresConfig);
+            namespace.addFunctions(operatorAndFunctionMap.get(false));
+            functionNamespaces = ImmutableMap.<String, FunctionNamespace>builder()
+                    .putAll(functionNamespaces)
+                    .put(catalog, namespace)
+                    .build();
+        }
+        else {
+            functionNamespaces.get(catalog).addFunctions(operatorAndFunctionMap.get(false));
+        }
+    }
+
+    public List<SqlFunction> list(Session session)
+    {
+        ImmutableList.Builder<SqlFunction> builder = ImmutableList.builder();
+        builder.addAll(operatorNamespace.list());
+        for (SqlPathElement element : session.getPath().getParsedPath()) {
+            FunctionNamespace namespace = functionNamespaces.get(getCatalog(element, session));
+            if (namespace != null) {
+                builder.addAll(namespace.list());
+            }
+        }
+        return builder.build();
+    }
+
+    public Signature resolveFunction(Session session, QualifiedName name, List<TypeSignatureProvider> parameterTypes)
+    {
+        try {
+            return operatorNamespace.resolveFunction(name, parameterTypes);
+        }
+        catch (Exception e) {
+            //do nothing and try to resolve the function using the path
+        }
+        for (SqlPathElement element : session.getPath().getParsedPath()) {
+            FunctionNamespace namespace = functionNamespaces.get(getCatalog(element, session));
+            if (namespace != null) {
```

Comment:
If we check against null and continue, that would help us get rid of the nestedness below:

```
if (namespace == null) {
    continue;
}
try {
    Signature signature = namespace.resolveFunction(name, parameterTypes);
    return qualifySignature(signature, getCatalog(element, session), TEMP_DEFAULT_SCHEMA);
}
catch (PrestoException e) {
    if (!e.getErrorCode().equals(FUNCTION_NOT_FOUND.toErrorCode())) {
        throw e;
    }
}
```

Reply:
I don't really like how the code looks with handling unresolved functions through exceptions. It would look nicer if the `namespace.resolveFunction` would just return `null` and only FunctionManager would throw.